### PR TITLE
419 new threshold bbs

### LINKED ISSUE #419: new Threshold BBs: AABB, k-Center, nearest neighbor

### PR DESCRIPTION
**Pull request by alkor200**
added all three new multidimensional threshold BBs

closes #419

### CODE REVIEW

File: algorithms/features/impl/k_center.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+                    return False
+        else:
+            return None
+
+    def _find_k_centers(self):
+        """
+        greedy algorithm that finds the k centers in datapoints
+        """
+        n = len(self._distance_matrix)
+        dist = [0] * n
```
Comment (tinsaye) [inline algorithms/features/impl/k_center.py:114]: Could this be done in one step with `[10**9] * n` ?

File: algorithms/features/impl/k_center.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        for i in range(n):
+            dist[i] = 10 ** 9
+
+        max_index = 0
+        for i in tqdm(range(self._k), desc="Calculating centers".rjust(27)):
```
Comment (tinsaye) [inline algorithms/features/impl/k_center.py:119]: Was searching the use of `i`, would be more obvious if we used `_` instead

File: algorithms/ids_k_center_main.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+from dataloader.direction import Direction
+
+from algorithms.ids import IDS
+
+from algorithms.features.impl.int_embedding import IntEmbedding
+from algorithms.features.impl.syscall_name import SyscallName
+from algorithms.features.impl.ngram import Ngram
+
+
```
Comment (tinsaye) [inline algorithms/ids_k_center_main.py:19]: Too many new lines :)

File: algorithms/ids_k_center_main.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        try:
+            LID_DS_BASE_PATH = os.environ['LID_DS_BASE']
+        except KeyError as exc:
+            raise ValueError("No LID-DS Base Path given."
+                             "Please specify as argument or set Environment Variable "
+                             "$LID_DS_BASE") from exc
+
+    LID_DS_VERSION = "LID-DS-2019"
+    SCENARIO_NAME = "CVE-2017-7529"
+
```
Comment (tinsaye) [inline algorithms/ids_k_center_main.py:37]: Too many new lines :)

File: algorithms/ids_k_center_main.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+    scenario_path = f"{LID_DS_BASE_PATH}/{LID_DS_VERSION}/{SCENARIO_NAME}"
+    dataloader = dataloader_factory(scenario_path, direction=Direction.BOTH)
+
```
Comment (tinsaye) [inline algorithms/ids_k_center_main.py:41]: Too many new lines :)

File: algorithms/ids_aabb_main.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    ### building blocks
+    syscall_name = SyscallName()
+    int_embedding = IntEmbedding(syscall_name)
+    ngram = Ngram([int_embedding], True, 5)
+    aabb = AABB(ngram)
+
+    ids = IDS(data_loader=dataloader,
+              resulting_building_block=aabb,
+              create_alarms=True,
+              plot_switch=False)
```
Comment (tinsaye) [inline algorithms/ids_aabb_main.py:49]: This will clash with @boneymoy pr, just as a reminder

**Review by tinsaye — COMMENTED** [review]
Looks good :+1:  just small questions

**Comment by alkor200** [comment]
thanks for the feedback @tinsaye 
all points were absolutely valid and i committed the changes :)

**Review by tinsaye — APPROVED** [review]
LGTM :)

**Pull request merged**